helix-core/src/main/java/org/apache/helix/controller/GenericHelixController.java

```diff
@@ -1072,6 +1146,15 @@ protected void checkLiveInstancesObservation(List<LiveInstance> liveInstances,
         }
       }
 
+      List<String> customizedStates = getEnabledCustomizedStates(manager);
+      for (String customizedState: customizedStates) {
+        for (String instance : curInstances.keySet()) {
+          if (lastInstances == null || !lastInstances.containsKey(instance)) {
+            addCustomizedStateListeners(manager, customizedState, instance);
```

Comment:
When you have add, we should have a corresponding remove to prevent listener leakage.